src/client/cl_main.cpp

```diff
@@ -4043,3 +4046,128 @@ void CL_GetVMGLConfig(vmglconfig_t *vmglconfig) {
 	vmglconfig->stereoEnabled = cls.glconfig.stereoEnabled;
 	vmglconfig->smpActive = cls.glconfig.smpActive;
 }
+
+void CL_ShaderStateChanged( void ) {
+	// Originally copied from CG_ShaderStateChanged, but rewritten to avoid issues of the original implementation and to
+	// support settings
+	char originalShader[MAX_QPATH];
+	char newShader[MAX_QPATH];
+	char settings[32];
+	char *timeOffset;
+	char *lightmapMode;
+	char *styleMode;
+
+	shaderRemapLightmapType_t lightmapModeValue;
+	shaderRemapStyleType_t styleModeValue;
+
+	const char *curPos = cl.gameState.stringData + cl.gameState.stringOffsets[ CS_MV_REMAPS ];
+	const char *endPos = curPos + strlen( curPos );
+
+	// Check configstring for prefix to avoid misinterpreting configstrings from mods
+	if ( !curPos || !*curPos || strncmp(curPos, "mvremap:", 8) ) return;
+
+	// Handling of these is really ugly. I originally wanted to handle them like the base game/cgame modules do, but
+	// those are prone to injections. Summary of delimeter issues:
+	//  - base uses '=', ':' an '@', all of them can be used in shader names
+	//  - .shader files don't seem to support spaces in shader names, but texture paths support spaces
+	//  - .shader files allow most other characters, including backspace '\'
+	//  -> using a delimeter is not a reliable option
+
+	// The next best thing seemed to be to give the length of the shader before its name to allow all characters to be
+	// used. As I don't want to waste any space (configstrings containing shader paths are wasteful as it is), I decided
+	// to encode the length in a single byte (MAX_QPATH is 80, so a single byte can easily hold the length). However the
```

Comment:
The `MAX_GAMESTATE_CHARS` define is currently set to `16000`. That means the sum of all configstrings may not exceed this value or clients are going to drop from the server. A single configstring may not exceed `8192` characters.

The shader remap example from the screenshots is already `3104` characters as it is and the server I was testing it on had around `9400` characters of configstrings set in total. As configstrings are part of the initial gamestate clients are unable to connect to a server long before hitting the maximum (due to the baseline being part of the gamestate as well). For that reason some servers have already begun to only sent vanilla remaps and userinfo configstrings once a client has fully connected. For that reason I would say every byte matters and we should actually announce whether clients support advanced remaps or not as part of their userinfo to allow servers to only send them if a client actually supports them (rather than servers sending both types).

Using ids sounds nice, but server and client would need to have the same ids for the same shaders. Theoretically that's what configstrings are used for: assigning ids to model paths, etc. to have a shared id for entities. So to share ids in advance we would need another system like configstrings before configstrings. That's obviously not an option here. Using hashes sounds like a nice option, but it would need to be collision free and collision free hashes usually come with a certain size/length as cost. With remaps sometimes being as simple as mapping `console` to `clear` a collision free hash would actually increase sizes in some cases.

For those reasons I would like to keep this as it is, but we should generally consider if we maybe want to add a similar system to configstrings, but without the limits as additional way to set certain values for future features after jk2mv 1.5.1 - exposing such a system to modules could actually be useful for mods, too.

Reply:
ok maybe it's not so complicated after all, there is just this extra long comment exploring all options that weren't used. Let's just keep it short and maybe leave single line explaining that we shift by X and it gives us certain range that will be valid for netchan, ok? Or even better - add an assert on path string length and comment the assert that it's for netchan - I always prefer to comment by code/asserts as much as possible because code actually gets tested and comments do not.

It made me confused because I thought the solution is super complicatedwhile reading it, but when I moved to actual code it turned out to be just a simple shift in the end. Once it's commited we're not changing it so what's the point describing whole thought process that went into it?

Reply:
I would prefer to keep the comment exploring the options around. It also hints at issues with the vanilla encoding and gives the reason behind the current implementation. We may not change it later, but someone might still wonder about it or fork/port some of the code.

Reply:
well, let's agree to disagree :)